pkg/metric/writer_prometheus.go

```diff
@@ -143,133 +146,155 @@ func (w *prometheusWriter) buildHelp(data *Datum) string {
+	if errors.As(err, are) {
+		return are.ExistingCollector.(T), nil
+	} else {
+		return *new(T), err
+	}
+}
 
-	metricI, ok := w.promMetrics.Load(id)
-	if !ok {
-		var err error
-		metric := w.createCounter(datum)
+func (w *prometheusWriter) promCounter(datum *Datum) {
+	metric := w.createCounter(datum)
 
-		err = w.addMetric(id, metric)
+	if err := w.registry.Register(metric); err != nil {
+		metricR, err := handleRegistrationError[*prometheus.CounterVec](err)
 		if err != nil {
-			return // error is logged in w.addMetric already
+			w.logger.Error(errFailedToRegisterMetricMsg, datum.MetricName, err)
+
+			return
 		}
 
-		metricI = metric
+		metric = metricR
+	} else {
+		err = w.addMetric()
+		if err != nil {
+			w.logger.Error(errFailedToAddMetricToPrometheusCounter, err)
+
+			return
+		}
 	}
 
-	metric := metricI.(*prometheus.CounterVec)
 	metric.
 		With(prometheus.Labels(datum.Dimensions)).
 		Add(datum.Value)
 }
 
 func (w *prometheusWriter) promGauge(datum *Datum) {
-	id := w.DatumId(datum)
-
-	metricI, ok := w.promMetrics.Load(id)
-	if !ok {
-		var err error
-		metric := w.createGauge(datum)
+	metric := w.createGauge(datum)
 
-		err = w.addMetric(id, metric)
+	if err := w.registry.Register(metric); err != nil {
+		metricR, err := handleRegistrationError[*prometheus.GaugeVec](err)
 		if err != nil {
-			return // error is logged in w.addMetric already
+			w.logger.Error(errFailedToRegisterMetricMsg, datum.MetricName, err)
+
+			return
 		}
 
-		metricI = metric
+		metric = metricR
+	} else {
+		err = w.addMetric()
+		if err != nil {
+			w.logger.Error(errFailedToAddMetricToPrometheusCounter, err)
+
+			return
+		}
 	}
 
-	metric := metricI.(*prometheus.GaugeVec)
 	metric.
 		With(prometheus.Labels(datum.Dimensions)).
 		Set(datum.Value)
 }
 
 func (w *prometheusWriter) promSummary(datum *Datum) {
-	id := w.DatumId(datum)
-
-	metricI, ok := w.promMetrics.Load(id)
-	if !ok {
-		var err error
-		metric := w.createSummary(datum)
+	metric := w.createSummary(datum)
 
-		err = w.addMetric(id, metric)
+	if err := w.registry.Register(metric); err != nil {
+		metricR, err := handleRegistrationError[*prometheus.SummaryVec](err)
 		if err != nil {
-			return // error is logged in w.addMetric already
+			w.logger.Error(errFailedToRegisterMetricMsg, datum.MetricName, err)
+
+			return
 		}
 
-		metricI = metric
+		metric = metricR
+	} else {
+		err = w.addMetric()
+		if err != nil {
+			w.logger.Error(errFailedToAddMetricToPrometheusCounter, err)
+
+			return
+		}
 	}
 
-	metric := metricI.(*prometheus.SummaryVec)
 	metric.
 		With(prometheus.Labels(datum.Dimensions)).
 		Observe(datum.Value)
 }
 
 func (w *prometheusWriter) promHistogram(datum *Datum) {
-	id := w.DatumId(datum)
-
-	metricI, ok := w.promMetrics.Load(id)
-	if !ok {
-		var err error
-		metric := w.createHistogram(datum)
+	metric := w.createHistogram(datum)
 
-		err = w.addMetric(id, metric)
+	if err := w.registry.Register(metric); err != nil {
```

Comment:
this code block looks like copy and paste.. would it be possible to abstract it in a better way?